### ISSUE TITLE
datastore: local test harness against emulators

### ISSUE THREAD

**Issue by mwitkow** [GoogleCloudPlatform/google-cloud-go#224]
We're using `stretchr/testify/suite` to test our code related to Datastore entities. For the purpose of these suites we wrote a testing harness around the old `gcd-v1beta2-rev1-4.0.0` Local Development server. 
Basically what it would do is: 
- initialize a random project name
- apply the user specified `index.xml` file
- start the server at a free, random port (so we can run these in parallel on Jenkins)
- stop the service and cleanup after the `suite` finished
  The `harness` would expose the new project name and a 

``` go
func (s *LocalDatabase) DatastoreClientOption() []cloud.ClientOption {
    return []cloud.ClientOption{
        cloud.WithEndpoint(fmt.Sprintf("http://localhost:%d/datastore/v1beta2/datasets/", s.serverPort)),
        cloud.WithScopes(),
        cloud.WithTokenSource(nil),
    }
```

that we'd use for our client connections. 

However, with Emulators we can no longer do this, since it is really really hard to bootstrap the `gcloud emulators` programatically. 

Is providing easy to use test bindings against Emulators something that you're planning?

**Comment by okdave** [comment]
In the short term you'll need to keep using the [standalone gcd emulator](http://storage.googleapis.com/gcd/tools/gcd-grpc-1.0.0.zip), since the gcloud one currently doesn't support gRPC. The easiest way to set up the storage package to communicate with that emulator is to set the environment variable `DATASTORE_EMULATOR_HOST`, but I am hoping to introduce a new `cloud.ClientOption` to make that simpler.

Though, it sounds like the main issue you're facing is the difficulty of scripting the gcloud-based emulator itself, is that right? Ideally, you'd like to be able to start one-or-more emulators from within your Go test itself?

**Comment by mbrevoort** [comment]
@okdave the standalone gcd emulator doesn't seem to support specifying a host to bind to which makes it tough to run inside of a container because it binds to `localhost` within the container and isn't accessible even with mapped ports 😞

**Comment by lgarron** [comment]
I would also find it useful to have *some* easy way to spin up a datastore emulator locally from Go tests.

https://github.com/chromium/hstspreload.appspot.com doesn't have any test yet because I still haven't found a good way to do it.
I can't get [`gcloud beta emulators datastore`](https://cloud.google.com/datastore/docs/tools/datastore-emulator) to work, and the standalone `gcd` emulator is really clunky, especially if the entire rest of your project works with `go get`. (However, unlike @mbrevoort I'm finding that the standalone emulator allows specifying a port).

**Comment by mbrevoort** [comment]
It's not that you can't specify the port, it's that you can no longer specify the host/ip to bind to rather than "localhost". FWIW I solved my problem, running inside a Docker container, by also starting a TCP proxy that binds to the external interface: https://github.com/mbrevoort/gcd-docker/blob/master/start.sh

**Comment by broady** [comment]
Actually, you can specify the host/port using the `DATASTORE_EMULATOR_HOST` environment variable.

https://github.com/GoogleCloudPlatform/gcloud-golang/blob/master/datastore/datastore.go#L63

**Comment by okdave** [comment]
I would like to make it easier to specify a local emulator (without resorting to env variables) through the options params. Something like `cloud.WithEmulator`?

I don't know what plans there are to make it easier to control the emulator from within Go itself. I definitely see the appeal of being able to bring up/down as many emulators as you like during a unit test. @pcostell might know what plans there are in this space?

**Comment by pcostell** [comment]
The emulator has some support for control, in particular:

GET request to "/" --> returns 200 and content "Ok" once the emulator is serving.
POST request to "/reset" --> clears all entities from the datastore. This is only valid if you aren't storing data to disk.
POST request to "/shutdown" --> shuts down the emulator cleanly, writing to disk if necessary.

**Comment by pcostell** [comment]
And regards to specifying the host, we'll support that in the next gcd release. (along with integration into gcloud).

**Comment by lgarron** [comment]
> POST request to "/reset" --> clears all entities from the datastore. This is only valid if you aren't storing data to disk.

Unfortunately, Go tests are meant to be run concurrently, and I have a need to test the output of a function that returns the entire database. Simply clearing the database isn't enough to ensure deterministic test results; I need independent test servers.

I've spent the last two days writing an ugly workaround (using `CloudDatastore.jar` directly rather than the `gcd.sh` wrapper) similar to the one described in the first post:

https://github.com/chromium/hstspreload.appspot.com/blob/5502fda4da88b653ccc461ba5ae3a194bab884d3/database/gcd/gcd.go

I've managed to get it to the point where different tests/goroutines can easily and safely run code on different local emulators at the same time.

    db, shutdown, err := NewLocalBackend()
    if err != nil {
      fmt.Printf("%s", err)
      return err
    }
    defer shutdown()

    // Create a context like normal.
    ctx, cancel := context.WithTimeout(context.Background(), 10 * time.Second)
    defer cancel()

    // Use `db` to construct the client.
    client, err := db.NewClient(c)
    if err != nil {
      return err
    }

However, it seems a bit silly to spend two days rewriting my code and inventing a new backend abstraction when I can normally just dive in and write tests for all my other Go code.

**Comment by derekperkins** [comment]
Is it too slow to just run your tests serially?

**Comment by lgarron** [comment]
> Is it too slow to just run your tests serially?

My impression is that sequential tests are an [antipattern](http://stackoverflow.com/questions/24375966/does-go-test-run-unit-tests-concurrently).

I also expect to add some end-to-end tests which take several seconds each (because they perform scans of live domains on the internet), and I run my tests as a pre-commit hook. It would be nice for those tests not to be inherently sequential.

If there was straightforward way to spawn a datastore test serve/call a mock server, perhaps there could be a reasonable convention to call sequential datastore. But since that's not possible, I prefer to write normal, isolated tests for now.

**Comment by lgarron** [comment]
> If there was straightforward way to spawn a datastore test serve/call a mock server,

In fact, the fact that I can spawn the servers in parallel is a by-product of how I structured my code. Setting up an emulator spawner in the first place took most of the work.

**Comment by derekperkins** [comment]
> My impression is that sequential tests are an antipattern.

There's a difference between serial and sequential. If the order of your tests matters, that's definitely an anti-pattern. We have a similar case with MySQL where all the tests can't be accessing the test server in parallel because they stomp on each other. We just run our tests one at a time, resetting the db before each test. 

It would be great if they could run in parallel, but they still run fast enough that it doesn't matter. Most of the time locally we aren't running the entire test suite, just the relevant tests. The whole test suite only runs pre-commit or on the CI server, so if it runs in 5 seconds or 50 seconds, it isn't a major drag. I just wouldn't spend a lot of time on it, though it sounds like you're about done. :)

> If there was straightforward way to spawn a datastore test serve/call a mock server

With the App Engine version of datastore, there is an actual command to spin up a mock server. https://cloud.google.com/appengine/docs/go/tools/localunittesting/reference#NewInstance I haven't used it in a while, but that actually took ~10 seconds to load each time, so we just wrapped our datastore calls so we could overwrite `Put` and `Get` requests per test.

**Comment by lgarron** [comment]
> We have a similar case with MySQL where all the tests can't be accessing the test server in parallel because they stomp on each other. We just run our tests one at a time, resetting the db before each test.

Is there a good place I can read up on this pattern for Go testing?

> With the App Engine version of datastore, there is an actual command to spin up a mock server.

Yeah, from the looks of it, `aetest` handles this kind of stuff very nicely. But I can't run directly on App Engine, so it's very hard to find any documentation on how I should test my datastore code.

**Comment by derekperkins** [comment]
This runs tests one at a time `go test -cover -p=1 ./...`

**Comment by lgarron** [comment]
> This runs tests one at a time go test -cover -p=1 ./...

So if someone just runs `go test`, your tests will fail for them? I guess that's okay if you can control the command people use for testing, but that seems a bit fragile to me. :-/

**Comment by okdave** [comment]
@lgarron Your test functions will only run one at a time unless you specifically opt into parallelism with `t.Parallel()` (see https://golang.org/pkg/testing/#T.Parallel). In other words, you don't have to worry about this unless you really want them to run at the same time.

**Comment by derekperkins** [comment]
@lgarron We built a function into our testing code that wipes and inserts our test data before each test iteration. It runs in < 1 ms, so it doesn't overly slow things down.

```
	for _, test := range tests {
		tester.ClearContext(c)
		if err := tester.CleanSlate(c, mysql.GlobalDB, apitest.CleanSlate); err != nil {
			t.Fatalf("tester.CleanSlate fail: %s", err)
		}

		// Run your tests here
	}
```

**Comment by derekperkins** [comment]
> Your test functions will only run one at a time unless you specifically opt into parallelism

@okdave That's not entirely true if you run `go test ./...`, at least as of a few versions ago. We got bit by non-deterministic tests and it took a while to track down.

**Comment by okdave** [comment]
@derekperkins sorry, I misunderstood the concern. The tests within a single package/binary will be run one at a time unless you call `t.Parellel`. However, `go test` will run multiple test binaries at once. So yes, you do need to be careful about inter-package interference.

**Comment by lgarron** [comment]
> POST request to "/reset" --> clears all entities from the datastore. This is only valid if you aren't storing data to disk.

I just tried this with `gcd.sh`/`CloudDatastore.jar`. The `/reset` endpoint returns a 404. :-(

**Comment by lgarron** [comment]
> I just tried this with gcd.sh/CloudDatastore.jar. The /reset endpoint returns a 404. :-(

Oops, I was using `GET`. However, `POST` seems to work. Sorry for the nose.

**Comment by derekperkins** [comment]
> So if someone just runs go test, your tests will fail for them? I guess that's okay if you can control the command people use for testing, but that seems a bit fragile to me.

It's not ideal, but it's certainly not more fragile than requiring a direct `datastore.jar` hack. I do look forward to this package providing a better way to spin up independent stores.

**Comment by lgarron** [comment]
> However, go test will run multiple test binaries at once.

Which I do, but I already handled that by asking the OS for an unused port in each program. :-)

> It's not ideal, but it's certainly not more fragile than requiring a direct datastore.jar hack.

Dunno. The alternative is relying on an externally launched datastore emulator, right?
(And that still doesn't handle if `go test` does spawn multiple test binaries in parallel.)

> I do look forward to this package providing a better way to spin up independent stores.

Indeed. :-)

**Comment by lgarron** [comment]
> for _, test := range tests {

I presume this means overriding [`MainStart`](https://golang.org/pkg/testing/#MainStart)?
I'm not brave enough to ignore the "not meant to be called directly", but knowing that there is no straightforward solution makes me comfortable with such a workaround.

I think I can settle for a shared lock that allows only one test to run at a time. Then each test can reset the database and run without spawning a new emulator, which sounds good enough.

**Comment by mbrevoort** [comment]
In case it helps, I run Go tests in parallel against the emulator but bootstrap the client with a random namespace for each test. Also, I only integration test the services that directly integrate with the datastore and mock away the datastore calls for all other tests.

**Comment by derekperkins** [comment]
> for _, test := range tests {
> I presume this means overriding MainStart?

@lgarron Apologies for not being clearer, that's for my table testing inside a single Go test function. Everything I am doing is standard `go test`. Here is a fuller contrived example.

```
func TestAdd(t *testing.T) {
	var tests = []struct {
		X  int
		Y  int
		ExpAns int
	}{
		{ 1, 2, 3 },
		{ 100, 55, 155 },
	}
    for _, test := range tests {
        tester.ClearContext(c)
        if err := tester.CleanSlate(c, mysql.GlobalDB, tester.CleanSlate); err != nil {
            t.Fatalf("tester.CleanSlate fail: %s", err)
        }

        // Run your tests here
        ans := Add(test.X, test.Y)
        if ans != test.ExpAns {
            t.Fatalf("Add fail: Expected: %d | Got: %d", test.ExpAns, ans)
        }
    }
}
```

**Comment by derekperkins** [comment]
As @okdave mentioned, if you don't explicitly put `t.Parallel()` in your test functions, they do not run concurrently inside of a single package. There isn't a shared lock to worry about, it just simply works. If you want to run on multiple packages with `go test ./...`, there's already a built in flag `-p=1` that ensures that your tests run one at a time. You shouldn't be coding anything to get that functionality.

> I run Go tests in parallel against the emulator but bootstrap the client with a random namespace for each test

@mbrevoort This is the best solution. 👏

**Comment by adams-sarah** [comment]
Closing for now due to inactivity. Feel free to reopen.

**Comment by lgarron** [comment]
Uh, I can't reopen, since I didn't file this bug.

But the lack of a Go API to connect to an emulator still makes it hard to test properly.

**Comment by yipjustin** [comment]
+1.

My team's test infrastructure has to use datastore directly in our
end-to-end test. Both the network overhead and cost are not negligible.

Would be good if we can use emulator directly.

**Comment by jba** [comment]
Moved to https://github.com/GoogleCloudPlatform/google-cloud-go/wiki/Feature-Backlog.

**Comment by zaddok** [comment]
The url in the above post doesn't seem to mention this issue, so what changed? What is the best practice to spin up a datastore emulator at the moment?

**Comment by jadekler** [comment]
@zaddok Please post the question as a new issue.
